### PR TITLE
[SPARK-4014] Add TaskContext.attemptNumber and deprecate TaskContext.attemptId

### PR DESCRIPTION
**Pull request by JoshRosen**
`TaskContext.attemptId` is misleadingly-named, since it currently returns a taskId, which uniquely identifies a particular task attempt within a particular SparkContext, instead of an attempt number, which conveys how many times a task has been attempted.

This patch deprecates `TaskContext.attemptId` and add `TaskContext.taskId` and `TaskContext.attemptNumber` fields.  Prior to this change, it was impossible to determine whether a task was being re-attempted (or was a speculative copy), which made it difficult to write unit tests for tasks that fail on early attempts or speculative tasks that complete faster than original tasks.

Earlier versions of the TaskContext docs suggest that `attemptId` behaves like `attemptNumber`, so there's an argument to be made in favor of changing this method's implementation.  Since we've decided against making that change in maintenance branches, I think it's simpler to add better-named methods and retain the old behavior for `attemptId`; if `attemptId` behaved differently in different branches, then this would cause confusing build-breaks when backporting regression tests that rely on the new `attemptId` behavior.

Most of this patch is fairly straightforward, but there is a bit of trickiness related to Mesos tasks: since there's no field in MesosTaskInfo to encode the attemptId, I packed it into the `data` field alongside the task binary.

### CODE REVIEW

**Comment by JoshRosen** [comment]
Maybe that test was flaky; let's see if it passes again (it'll retest since I pushed a commit to fix a merge conflict).

I've updated this patch to not modify `attemptId` but to introduce `attemptNumber` and deprecate `attemptId`.  I think it will be confusing to have `attemptId` have different behavior in different branches, especially since it seems like functionality that might be nice to rely on when writing certain types of regression tests.  Since this patch doesn't change any behavior, I'd like to backport it to maintenance branches so that we can rely on it in test code.  If we decide to do that, the committer should update the MiMa exclusions on cherry-pick.

**Comment by JoshRosen** [comment]
@pwendell I've updated this based on our previous discussions; please take another look.

**Comment by pwendell** [comment]
@joshrosen LGTM relating the renaming.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala [hunk 1]
```diff
@@ -290,13 +291,22 @@ private[spark] class MesosSchedulerBackend(
       .setType(Value.Type.SCALAR)
       .setScalar(Value.Scalar.newBuilder().setValue(scheduler.CPUS_PER_TASK).build())
       .build()
+    // Encode the attemptId as part of the data payload, since there's not a MesosTaskInfo field
+    // to hold it:
+    val data: ByteString = {
+      val serializedTask = task.serializedTask
+      val dataBuffer = ByteBuffer.allocate(4 + serializedTask.limit())
+      dataBuffer.putInt(task.attemptNumber)
+      dataBuffer.put(serializedTask)
```
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala]: Can we introduce another wrapper here instead? I'm imagine we will be adding more fields to serialize to Mesos executors, and it's a lot easier to maintain a struct then position with types and offsets.

Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala]: That's a good idea; putting the serialization / deserialization code in the same wrapper class will make it much easier to verify that it's correct / test it separately.  I'll push a new commit to do this.

**Comment by tnachen** [comment]
Thanks for adding the wrapper, the Mesos portions looks good to me.

**Comment by JoshRosen** [comment]
Noo!  This became merge-conflicted.  Let me bring it up to date...

**Comment by JoshRosen** [comment]
I'm going to merge this into `master` (1.3.0) since it's a blocker for some tests that I want to write.  I'll look into backporting this into maintenance branches, too, since that would allow me to backport regression tests that use the new methods.

**Pull request closed without merging**